### ISSUE TITLE
Avoid double loading of python modules.

### ISSUE THREAD

**Issue by deathaxe** [jisaacks/GitGutter#396]
Sublime Text actively loads all modules located in a package's root directory to search for API classes to hook. If such a module is imported by another one _(see: git_gutter.py)_ it will be loaded again by the python interpreter. All classes and globals are initialized twice, which is a waste of time and may cause issues in some cases (e.g.: settings) as multiple instances are created for each python object.

This issue can easily be avoided by moving all modules which are imported by at least one module into a folder. I already started to do move some python sources into the `modules` sub directory.

In the end all modules are imported by `git_gutter.py` need to be moved to avoid double loading.

In order not to split existing modules like git_gutter_popup.py, which may include both, the implementation and API I suggest the following folder structure.

```
[modules]
   - __init__.py    # exports API classes
   - events.py      # git_gutter_events.py
   - commands.py    # git_gutter.py
   - compare.py     # git_gutter_compare.py
   - goto.py        # git_gutter_jump_to_changes.py
   - handler.py     # git_gutter_handler.py
   - popup.py       # git_gutter_popup.py
   - settings.py    # git_gutter_settings.py
   - promise.py
   - show_diff.py   # git_gutter_show_diff.py

- plugin.py
```

The `git_gutter_` prefix of the module names was omitted as shorter is better to read and it seems unnecessary with the new structure.

This structure enables us to use relative imports on both ST2 and ST3 within the modules which makes all the `try..except..` stuff unnecessary for module imports.

The `plugin.py` simply imports the `modules` folder and `__all__` exported classes and functions. ST will see this file only, which results in the following console output upon startup.

```
reloading plugin GitGutter.plugin
```

To get an idea of how the new structure would look like, you can have a look on PR #385. It comes with the basic infrastructure required to handle this new folder/module structure.

I'd like to know if this is a structure you can agree with.

**Comment by r-stein** [comment]
Have you tried some time measurement to check whether it is really faster?

I am quite neutral to it, because currently every plugin is tighten together anyway. I see as an additional advantage, that you can conditionally load plugins, e.g. load popup only in ST3. Disadvantage would be, that the files are not automatically reloaded by saving during the development. In addition it may result in some problem on updating if the function signature changed and the files are not reloaded correct.

However we should:
- pipe `plugin_loaded`/`plugin_unloaded` to the subfiles
- not abuse the `__all__` import statement, but use something else

**Comment by deathaxe** [comment]
### Benchmark

I did not manage to do some scientific measurements on timings, which even might be a bit tricky upon startup, but found each module being reloaded and printed to console to require a minimal set of time to be handled. 

I must confess this is not a primary factor to do this suggestion. I am worried more about the problem of multiple instances of all the created python objects due to the double loading bug.

### Automatic package reloading

In fact files in sub modules are not reloaded automatically, but  Sublime Text 3 does not reload packages from root folder which use asynchronous API well, too. I experienced issues with ST's package reloading several times anyway. Everything works fine as long as a package uses synchronous api only.

This was in fact one reason to do some more investigation on package loading process. Another reason was the settings module which @rchl struggled with last year. Those issues with multiple instances were caused by the double loading bug.

#### 1. Developement

_During development_ the [AutomaticPackageReloader](https://packagecontrol.io/packages/AutomaticPackageReloader) can easily be used to reload all sub modules. I created a key binding `Ctrl+F5` to reload sub packages, but you can ask it to automatically reload packages on save. It works very well - even better as ST3's default mechanism.

#### 2. Package Upgrade

I read about possible issues during upgrade as well as python won't reload modules on its own, if they already exist in the cache. Therefore I had a look on how AutomaticPackageReloader works. It uses a modified version of GitSavvy's reloader.py to force reloading all sub modules in exactly the same order the original import run would do. It ensures not to create multiple instances of existing modules. It's some kind of magic, but works very well.

The proposed structure moves all modules to the sub directory. As a result all changes on existing modules would take effect after restarting ST only, if we do nothing. We could run into trouble only, if a new module is added to GitGutter, which is loaded by ST after upgrading and which makes use of changed existing modules, which are not reloaded.

If we want to address this rare possible issue we could use a modified reloader.py from GitSavvy and _Package Control's_ event API to force reloading all sub modules after upgrading GitGutter. I played around with it already. This would require ca. 200 lines of code to handle all of it.

### Pipe plugin_loaded / plugin_unloaded

I would suggest to add those functions to the plugin.py in the main folder to let them be called by ST and explicitly add all required functions from sub modules there. This way the plugin.py is the one and only place where to handle module initialization. The `git_gutter_settings.py` is the only module at the moment, which makes use of `plugin_loaded()`. My work on fixing issue #184 will remove this function call as settings loading is handled somewhat different.  This results in none of the modules requiring those functions at all.

### Abuse `__all__`

Why do you think `__all__` is abused by this use? I found many modules even in the stdlib, which make use of this property to explicitly manage a list of API functions to export. As a ST package needs to export ST API classes only, I find this solution quite appropriate. (doesn't mean it is :-) )

The main intention of this solution is to avoid the horrible long `try ... except ...` structures to handle ST2 and ST3 module imports. The `__init__.py` as well as any other module can make use of a relative import on ST2 and ST3. The plugin.py would be the only module which requires the `try ... except ...` structure to handle the imports separately for ST2 and ST3 in 4 lines of code. 

The alternative is to explicitly import all required classes in the plugin.py directly in the same way the git_gutter.py does today. One long list of relative imports for ST3 in the `try` statement and one long list of absolute imports for ST2 in the `except` statement.

The plugin.py would look like:

```python
"""Load and Unload all GitGutter modules.

This module exports __all__ modules, which Sublime Text needs to know about.
"""

try:
    from .modules.events import EventListener
    from .modules.commands import (
        GitGutterCommand, GitGutterCompareBranchCommand,
        GitGutterCompareCachedCommand, GitGutterCompareCommitCommand,
        GitGutterCompareFileCommitCommand, GitGutterCompareHeadCommand,
        GitGutterCompareOriginCommand, GitGutterCompareTagCommand,
        GitGutterNextChangeCommand, GitGutterPrevChangeCommand,
        GitGutterShowCompareCommand)
    from .modules.popup import (
        GitGutterDiffPopupCommand, GitGutterReplaceTextCommand)
    from .modules.settings import (
        GitGutterEditSettingsCommand, GitGutterOpenFileCommand)
except ValueError:
    from modules.events import EventListener
    from modules.commands import (
        GitGutterCommand, GitGutterCompareBranchCommand,
        GitGutterCompareCachedCommand, GitGutterCompareCommitCommand,
        GitGutterCompareFileCommitCommand, GitGutterCompareHeadCommand,
        GitGutterCompareOriginCommand, GitGutterCompareTagCommand,
        GitGutterNextChangeCommand, GitGutterPrevChangeCommand,
        GitGutterShowCompareCommand)
    from modules.popup import (
        GitGutterDiffPopupCommand, GitGutterReplaceTextCommand)
    from modules.settings import (
        GitGutterEditSettingsCommand, GitGutterOpenFileCommand)
```

Nothing I would oppose to at all, but found the idea of using the `__all__` more elegant. Same number of code lines in the end. So ...

**Comment by deathaxe** [comment]
I managed to do some benchmarks by hacking the sublime_plugin.py with following results.

Loading time
1. master branch: 28ms
2. local restructured dev branch: 18ms

The difference might be even lager when comparing the same but only restructured code base. 

The local branch I used for (2) differs a lot from current master as it already makes use of some experimental features and changes. Even though it is somewhat larger it only needs 64% of the time to load compared to the current master branch. Ok, the absolute time might not be so much, but imagine 100 or 200 plugins with such a difference.
